### PR TITLE
Support Cirrus Logic CS42L42 headphone codec

### PR DESCRIPTION
**Pull request by brentlu**
It's a series to:
1. Add a machine driver, sof_cs42l42, to support Cirrus Logic CS42L42 + Maxim MAX98357A on GLK devices.
2. Move MAX98357A support to a common module, sof_maxim_common
3. Code refactory to sof_rt5682 and sof_cs42l42 to use sof_maxim_common to support Maxim MAX98357A

### CODE REVIEW

**Comment by brentlu** [comment]
The topology is here: https://github.com/thesofproject/sof/pull/4121

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: 2021

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c:100]: is the mapping correct? It seems that DA7219 had a quirk compared to the other codecs, the regular sequence is

````
	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
````

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: this looks like copy-paste, is this .ignore_pmdow_time necessary?

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c:440]: this looks like copy-paste, we don't have any other platform with cs32l42.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: "GPL"

also rename the 'SOF Audio Machine driver', it's too vague, you need a reference to cs42l42

File: sound/soc/intel/boards/Kconfig
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: we can simplify this, we don't have a case with BAYTRAIL.
@kv2019i can you double-check the HDaudio parts, I can't recall what's necessary.

File: sound/soc/intel/boards/sof_rt5682.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c:451]: comment on commit message: I think you meant 'refactor'

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c:509]: you can move the maxim common part first, as a preparation, and then squash all the cs42l42 changes. No need to show the changes in separate patches, you can just add the new machine driver in one patch, i'll be simpler to review.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: fixed

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c:100]: Yes, ODM validates the pause/volume up/volume down buttons.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: According to ODM test result, there is no pop/static noise after removing this option. So I remove it.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c:440]: Suggest keeping it since we are going to support ADL platform in the future.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: fixed

File: sound/soc/intel/boards/Kconfig
Comment (brentlu) [reply sound/soc/intel/boards/Kconfig]: removed

File: sound/soc/intel/boards/sof_rt5682.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c:451]: fixed

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c:509]: Rebase the commits to merge cs42l42 changes.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_cs42l42.c]: this can be `const`

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_cs42l42.c]: nitpick: if you do this right at the top of the function, maybe just in the initialisation, you can also use `jack` in `snd_soc_card_jack_new()` call above

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (lyakh) [inline sound/soc/intel/boards/sof_cs42l42.c]: braces shouldn't be used

**Review by lyakh — COMMENTED** [review]
some minor cosmetic improvement suggestions

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: fixed

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: fixed, thanks for the review~

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: done

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [reply sound/soc/intel/boards/sof_cs42l42.c:440]: ok

**Review by plbossart — APPROVED** [review]
thanks @brentlu, nice work.

**Comment by plbossart** [comment]
I had to manually approve @brentlu so let's wait until we see the GitHub Actions results + another approved to merge

**Review by plbossart — CHANGES_REQUESTED** [review]
I have to take my approval back after looking at the topology.
It's simpler if we have first the SSP links added, then dmic and HDMI last. The topology changes [1] can be simplified a lot by using the same order as for DA7219.

[1] https://github.com/thesofproject/sof/pull/4121/commits/ef0ebafad22e2cf46dc6bc40396d75d9590dd2f4#diff-ef984278da4b010127ccd588daa6d53c55bff7a852c2ef42f7b39a82d4bd3951R195

**Comment by brentlu** [comment]
Adjust the DAI Link sequence in sof_card_dai_links_create() function to follow the order: SPK, HEADPHONE, DMIC, HDMI

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
restarting SOF CI tests since we are missing the configuration for the new machine driver

**Comment by plbossart** [comment]
SOFCI TEST

**Review by plbossart — CHANGES_REQUESTED** [review]
@brentlu I added the missing config and CI reports errors with make W=1 and also checkpatch.pl --strict has quite a few warnings. Nothing critical but let's make this PR shiny for upstream. Thanks!

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_cs42l42.c:129]: nitpick: How about remove `clk_freq` and just use `3072000` here?

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (bardliao) [inline sound/soc/intel/boards/sof_cs42l42.c]: Sorry, I didn't get what is `i` for?

**Review by bardliao — COMMENTED** [review]
Looks quite good to me, just some minor comments.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: seems useless so i remove it. thanks~

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c:129]: we may switch to 2.4 MHz in the future so suggest keeping it.

**Comment by brentlu** [comment]
warnings from checkpatch are fixed

**Comment by brentlu** [comment]
False alarm?

-------------------------------------------------------------------
Commit 429ffb620dcd ("ASoC: Intel: add sof-cs42l42 machine driver")
-------------------------------------------------------------------
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#74: 
new file mode 100644

total: 0 errors, 1 warnings, 0 checks, 570 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Commit 429ffb620dcd ("ASoC: Intel: add sof-cs42l42 machine driver") has style problems, please review.

**Comment by brentlu** [comment]
@plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?

**Comment by plbossart** [comment]
> False alarm?
> 
> ## Commit [429ffb6](https://github.com/thesofproject/linux/commit/429ffb620dcdda38bde01b36be030e2fe52dc848) ("ASoC: Intel: add sof-cs42l42 machine driver")
> WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
> #74:
> new file mode 100644
> 
> total: 0 errors, 1 warnings, 0 checks, 570 lines checked
> 
> NOTE: For some of the reported defects, checkpatch may be able to
> mechanically convert to the typical style using --fix or --fix-inplace.
> 
> Commit [429ffb6](https://github.com/thesofproject/linux/commit/429ffb620dcdda38bde01b36be030e2fe52dc848) ("ASoC: Intel: add sof-cs42l42 machine driver") has style problems, please review.

yes, that's normal and not a problem.

**Comment by plbossart** [comment]
> @plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?

let me check with Mark Brown, there's a missing dependency upstream, some previous patches were not applied, and if you send the patches they will not apply directly. thanks!

**Comment by plbossart** [comment]
still a make W=1 issue @brentlu . You can fix this in your setup by running 'make sound/soc/codecs/ W=1' (the last / matters). Thanks!

**Comment by plbossart** [comment]
> > @plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?
> 
> let me check with Mark Brown, there's a missing dependency upstream, some previous patches were not applied, and if you send the patches they will not apply directly. thanks!

ok, so I need to resend the previous series which was partly missed for some reason, and I'll include your PR on top once all the warnings are removed. Does this work for you @brentlu? Thanks!

**Comment by brentlu** [comment]
The warning is from the codec header file which we don't need it. I've removed it from machine driver and send a mail to Cirrus team to fix it.

**Comment by brentlu** [comment]
Thanks. I think it's the best option.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: sorry, forgot this one. This was removed in an earlier commit in other platforms, for consistency please remove the .nonatomic here as well.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: remove

**Review by plbossart — CHANGES_REQUESTED** [review]
sorry, one more fix needed, the code was inspired by an earlier version of sof-rt5682 but the .nonatomic property is not needed.

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: removed

File: sound/soc/intel/boards/sof_cs42l42.c
Comment (brentlu) [reply sound/soc/intel/boards/sof_cs42l42.c]: remove, BTW, how do I tell if the flag is needed or not?

Comment (plbossart) [reply sound/soc/intel/boards/sof_cs42l42.c]: this is not needed for BE, it's only needed for FE, and FEs are now handled in the topology. You don't need to worry about it for newer machine drivers.

Comment (plbossart) [reply sound/soc/intel/boards/sof_cs42l42.c]: see commit 961cd002aacc9ff027f1eabdcd5b8cd0c35346a6 @brentlu

**Review by plbossart — APPROVED** [review]
let's merge, thanks @brentlu

**Comment by plbossart** [comment]
I'll use my superpowers to merge and avoid more delays

**Pull request merged**
